src/budget/budgetmanager.cpp

```diff
@@ -969,6 +978,30 @@ int CBudgetManager::ProcessProposalVote(CBudgetVote& vote, CNode* pfrom)
     }
 
     const CTxIn& voteVin = vote.GetVin();
+
+    // See if this vote was signed with a deterministic masternode
+    auto mnList = deterministicMNManager->GetListAtChainTip();
+    auto dmn = mnList.GetMNByCollateral(voteVin.prevout);
+    if (dmn) {
+        if (!vote.CheckSignature(dmn->pdmnState->keyIDVoting)) {
+            LogPrintf("mvote - signature invalid from dmn %s\n", dmn->proTxHash.ToString());
+            return 100;
+        }
+        AddSeenProposalVote(vote);
+        std::string strError = "masternode not valid or PoSe banned";
+        if (mnList.IsMNValid(dmn) && UpdateProposal(vote, pfrom, strError)) {
```

Comment:
Better to raise the DoS level if the DMN isn't valid (which atm can only happen if the collateral was spent), plus move the `AddSeenProposalVote` after the DMN validity check.
Otherwise a dead DMN could fill other peer's memory with invalid votes.

Reply:
Agreed. But better to address these things in a separate PR (possible multiple PRs, overhauling the budget code).
In this PR I tried to keep the behavior changes to a minimum, with the goal of integrating DMNs into the old system.